### PR TITLE
Remove complex queries from client.rb and project.rb

### PR DESCRIPTION
**Pull request by mohinid**
## Notion card
https://www.notion.so/saeloun/Move-Complex-Query-to-ES-bf5f01173ca045bb94c9bca6aa82e2fd

## Summary
- Remove "new_line_item_entries" method from client.rb and use generate_invoice controller to fetch line items.
- Remove "search_all_projects_by_name" method from project
- Rewrite internal_api/v1/project/search_controller.rb using searchkick to fetch searched projects(with string matching project name or client name)
- Additional Front End changes related to API calls.
- Update RSpecs

## Preview
Screenshot 2022-10-10 at 12 24 47 PM
Screenshot 2022-10-10 at 12 25 02 PM

## Type of change
Please delete options that are not relevant.
- [x] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Via Postman, Local and Rspec

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code
- [x] I have added automated tests for my code

### CODE REVIEW

**Comment by mohinid** [comment]
@akhilgkrishnan It's giving this error while accessing the project path. Please suggest.
Screenshot 2022-10-10 at 12 27 46 PM

**Comment by mohinid** [comment]
@prasanthchaduvula  As we have removed new_line_item_entries" method from client.rb, can you please check this file - 
Issue in edit invoice from frontend file- miru-web/app/javascript/src/components/Invoices/Edit/index.tsx

It's giving 404 after successfully loading the edit invoice page.

`// if (selectedClient.value != 0) {
    //   fetchNewLineItems(navigate, selectedClient.value, setSelectedClient, setLineItems);
    // }`

**Comment by mohinid** [comment]
@apoorv1316  please review.

**Review by apoorv1316 — CHANGES_REQUESTED** [review]
@mohinid Project search is not working and throwing an error. Please check https://www.loom.com/share/b760c47313f94ea59a33c03929ad9259

**Comment by mohinid** [comment]
> @mohinid Project search is not working and throwing an error. Please check https://www.loom.com/share/b760c47313f94ea59a33c03929ad9259

@apoorv1316  please do Project.reindex on your console before running tests. or do db reset.

**Comment by supriya3105** [comment]
@prasanthchaduvula Can you review and approve as well.

File: app/controllers/internal_api/v1/projects/search_controller.rb [hunk 1]
```diff
@@ -4,13 +4,30 @@ class InternalApi::V1::Projects::SearchController < InternalApi::V1::Application
...
   private
 
-    def search_params
-      params.require(:search_term)
+    def search_term
+      @search_term ||= (params[:search_term].present?) ? params[:search_term] : ""
+    end
+
+    def client_list
+      @_client_list ||= Company.find(current_company.id).clients.pluck(:id).uniq
```
Comment (shalapatil) [inline app/controllers/internal_api/v1/projects/search_controller.rb:21]: current_company.clients.pluck(:id).uniq

File: app/models/company.rb [hunk 2]
```diff
@@ -57,7 +57,7 @@ def project_list(client_id = nil, user_id = nil, billable = nil, search)
       project_list.each do |project|
         if id == project.id
           billable_array.push(project.is_billable)
-          client_name_array.push(project.client_name)
+          client_name_array.push(project.project_client_name)
```
Comment (shalapatil) [inline app/models/company.rb:60]: why this change?

File: app/models/project.rb [hunk 3]
```diff
@@ -40,6 +40,20 @@ class Project < ApplicationRecord
...
+
+  def search_data
+    {
+      id: id.to_i,
+      name:,
+      description:,
+      billable:,
+      client_id:,
+      client_name:
+    }
```
Comment (shalapatil) [inline app/models/project.rb:55]: you need to reindex project when client data changes. Add that code to client model

**Review by shalapatil — CHANGES_REQUESTED** [review]

File: app/models/company.rb (hunk 2 above)
Comment (mohinid) [reply app/models/company.rb:60]: hi this change was required as the client project list was failing. Akhil made this name change to resolve conflict.

File: app/models/project.rb (hunk 3 above)
Comment (mohinid) [reply app/models/project.rb:55]: sure

File: app/controllers/internal_api/v1/projects/search_controller.rb [hunk 4]
```diff
@@ -4,13 +4,30 @@ class InternalApi::V1::Projects::SearchController < InternalApi::V1::Application
...
+    end
+
+    def searched_projects
+      search_result = Project.search(
+        search_term,
+        fields: [:client_name, :name],
+        match: :text_middle,
+        where: { client_id: client_list },
+        includes: [:client]
```
[the remark this answers is not in the document]
Comment (mohinid) [reply app/controllers/internal_api/v1/projects/search_controller.rb:30]: yes, no need of username as user is not displayed on the projects tab.

File: app/controllers/internal_api/v1/projects/search_controller.rb (hunk 1 above)
Comment (mohinid) [reply app/controllers/internal_api/v1/projects/search_controller.rb:21]: added

**Comment by mohinid** [comment]
@shalapatil  thanks for your comments! Updated the code accordingly. Please review it now.

File: app/controllers/internal_api/v1/projects/search_controller.rb [hunk 5]
```diff
@@ -4,13 +4,30 @@ class InternalApi::V1::Projects::SearchController < InternalApi::V1::Application
   skip_after_action :verify_authorized, only: [:index]
 
   def index
-    projects = Project.search_all_projects_by_name(search_params, current_company.id)
-    render json: { projects: }, status: :ok
+    # projects = Project.search_all_projects_by_name(search_params, current_company.id)
+    render :index, locals: {
+      searched_projects:,
+      total_searched_projects: searched_projects.total_count
```
Comment (shalapatil) [inline app/controllers/internal_api/v1/projects/search_controller.rb:10]: as you are calling searched_projects twice, please either store it against var here or adding caching to the function

File: app/models/client.rb [hunk 6]
```diff
@@ -41,25 +41,12 @@ class Client < ApplicationRecord
...
-        "timesheet_entries.id as timesheet_entry_id,
-         users.first_name as first_name,
-         users.last_name as last_name,
-         timesheet_entries.work_date as date,
-         timesheet_entries.note as description,
-         project_members.hourly_rate as rate,
-         timesheet_entries.duration as quantity"
-      ).where.not(id: selected_entries).order("timesheet_entries.work_date").distinct
+  def reindex_projects
+    Project.reindex
```
Comment (shalapatil) [inline app/models/client.rb:49]: projects.reindex

File: app/models/project.rb [hunk 7]
```diff
@@ -40,6 +40,20 @@ class Project < ApplicationRecord
 
   # Callbacks
   after_discard :discard_project_members
+  delegate :name, to: :client, prefix: true, allow_nil: true
+
+  searchkick text_middle: [:name, :client_name]
```
Comment (shalapatil) [inline app/models/project.rb:45]: if we are specifying text_middle here, then do we really need to speicify it while doing the actual search query?

**Review by shalapatil — CHANGES_REQUESTED** [review]

File: app/models/project.rb (hunk 7 above)
Comment (mohinid) [reply app/models/project.rb:45]: yes, it's not working without explicitly mentioning it in both places. can show you from the console.

**Review by apoorv1316 — APPROVED** [review]

**Review by shalapatil — APPROVED** [review]

**Pull request merged**